File: python/767_重构字符串_medium.py
class Solution:
    def reorganizeString(self, S: str) -> str:
        length = len(S)
        if length<2:
            return S
        chCount = [0 for i in range(26)]
        maxCount = 0
        for ch in S:
            chCount[ord(ch)-ord('a')] += 1
            maxCount = max(maxCount, chCount[ord(ch)-ord('a')])
        if maxCount > (length+1)//2:
            return ""
        evenPos, oddPos = 0, 1
        halfCount = length//2
        ans = [' ' for i in range(length)]
        for i in range(26):
            ch = chr(ord('a')+i)
            while(chCount[i] and chCount[i] <= halfCount and oddPos < length):
                ans[oddPos] = ch
                oddPos += 2
                chCount[i] -= 1
            while(chCount[i]):
                ans[evenPos] = ch
                evenPos += 2
                chCount[i] -= 1
        return ''.join(ans)

File: cpp/767_重构字符串_medium.cpp
class Solution {
public:
    string reorganizeString(string S) {
        int length = S.length();
        if(length<2)
            return S;
        int maxCount = 0;
        vector<int> chCount(26, 0);
        for(int i=0; i<length; i++){
            chCount[S[i]-'a'] += 1;
            maxCount = max(maxCount, chCount[S[i]-'a']);
        }
        if(maxCount>(length+1)/2)
            return "";
        string temp(length, ' ');
        int evenPos = 0, oddPos = 1;
        int halfLength = length/2;
        for(int i=0; i<26; i++){
            char ch = (char)('a'+i);
            while(chCount[i] && chCount[i]<=halfLength && oddPos<length){
                temp[oddPos] = ch;
                oddPos += 2;
                chCount[i]--;
            }
            while(chCount[i]){
                temp[evenPos] = ch;
                evenPos += 2;
                chCount[i]--;
            }
        }
        return temp;
    }
};